Add synthetic diag: mag probes and flux loops

orso82 opened the pull request:
Opening this PR on behalf of @ggdose

ggdose committed:
Add synthetic diag: mag probes and flux loops

orso82 committed:
simplified and formatted flux_loops and magnetic_probes routines

orso82 committed:
silly formatter, I win

anchal-physics commented:
### add functions
Please add Functions to add flux_loop and magnetic_probe with doc strings like:

```julia


"""
    add_flux_loop!(
        config::Union{String, Dict{Symbol, Any}}=default_flux_loop,
        @nospecialize(ids::IMAS.dd)=IMAS.dd();
        overwrite::Bool=false,
    )::IMAS.dd

Add flux_loop to IMAS structure using a `JSON` file or Julia `Dict`.
"""

"""
    add_flux_loop!(
        config::Union{String, Dict{Symbol, Any}}=default_flux_loop,
        @nospecialize(ids::IMAS.dd)=IMAS.dd(),
        PSI_interpolant::Interpolations.AbstractInterpolation,;
        overwrite::Bool=false,
    )::IMAS.dd

Add flux_loop to IMAS structure using a `JSON` file or Julia `Dict` and compute
the measured poloidal flux using the provided PSI_interpolant.
"""

```

You can conveniently use `add_diagnostic!` function as used in `add_interferometer!`here: https://github.com/ProjectTorreyPines/FusionSyntheticDiagnostics.jl/blob/505699e6a72a499838847deb347c65a9938ae7ed/src/interferometer.jl#L26

Similarly, add a function for called `add_magnetics_probe!()`

These functions allow for an easy way to add test diagnostics to an existing `ids` or loop through locations of such diagnostics for optimization purposes. The file reading part makes it easy to store and share such diagnostic locations in the device, real or conceptual.

### Remove plural forms
Remove plural forms from compute function names to conform to the rest of the package and IMAS naming convention as well. So the functions should be called `compute_flux_loop!` and `compute_magnetic_probe!`

### Add tests

Please add a minimal working test in runtests. For this, it would be good to add some `default_flux_loop.json` and `default_magnetics_probe.json` file in src as well, just like the other diagnostics.

### (OPTIONAL)`b_field_pol_probe`
It might be better to call the functions `compute_b_field_pol_probe!` leaving room for a future `compute_b_field_tor_probe!` and making it easy to just follow IMAS naming conventions to remember add and compute function names. Maybe we can have something like:

`add_flux_loop!`
`add_b_field_pol_probe!`
`add_b_field_tor_probe!`
`add_rogowski_coil!`
`add_shunt!`

and

`add_magnetics!` to add all magnetics.

orso82 commented:
@anchal-physics I am not sure I understand the `add_flux_loop!()` functions. For what machine?

orso82 committed:
compat IMAS = "1.2.2, 2, 3, 4"

orso82 committed:
IMAS = "1.2.2, 2, 3, 4, 5"

orso82 commented:
@anchal-physics could you and @ggdose please work together to push this through the finish line?
If it helps, the magnetics (and more) for DIII-D are available here: https://github.com/ProjectTorreyPines/FUSE.jl/blob/master/sample/D3D_machine.json

anchal-physics committed:
Consolidated magnetic probes into single file

flux_loops.jl, magnetic_probes.jl -> magnetics.jl

old function names have been changes, now user just need to use

add_magnetics! to add name, identifier, and geometrical quantities
related to all magnetic diagnostics such as b_field_pol_probes,
b_field_tor_probes, flux_loops, rogowski_coil and shunt.

compute_magnetics!
is a single overloaded function to compute all magnetic diagnostics
given required quantities have been passed as keywords. Currently
supported diagnostics for computation are b_field_pol_probes and
flux_loops both of which require PSI_interpolant for calculation.

anchal-physics committed:
Minor format updates

anchal-physics committed:
Added docs for magnetics

anchal-physics commented:
I think this is ready to merge now. I've satisfied all of my own criteria. Please let me know if this looks to you.
@ggdose can you try this branch and if possible, add a MWE in the runtests.jl. If we add a test, it will be perfect and will also document a way to run the calculation.

anchal-physics committed:
Added test for add_magnetics!;Compute test remaining

anchal-physics committed:
Added tests for flux loops and pol field probes

anchal-physics commented:
This is ready to merge in my opinion. @ggdose please check if the everything looks ok.

orso82 commented:
thank you @ggdose and @anchal-physics for pushing this through the finish line!

anchal-physics committed:
Releasing 2.2.0 with magnetics

* b_field_pol_probe and flux_loop magnetics have been added
* These magnetic diagnostics are in magic implementation and do not include physical modeling
* They utilize the equilibrium profile to produce ideal results
